### PR TITLE
Upgrade to V2.0

### LINKED ISSUE #81: Upgrading to v2
Biscuit v2 was released [recently](https://www.biscuitsec.org/blog/biscuit-2-0/), so I'm looking at upgrading existing implementations.
#56 was already well advanced thanks to the work of @daeMOn63 and @titanous, apparently the only part missing was updating the parser?

I think the v2 upgrade could start directly from this PR, the parser update is not strictly necessary (it can be done in parallel).
As for v1, there's an [upgrade issue in the spec repository](https://github.com/biscuit-auth/biscuit/issues/72) detailing the various changes to implement, and for each of them, the samples and authorization results to test them.

I'm opening a PR to track that work: #82

I'd be happy to get some help on this, as I am not very familiar with Go :sweat_smile:

### PR DESCRIPTION
**Pull request by Geal**
fix #81

TODO V2 migration:
- [x] crypto update
- [x] sealed tokens
- [x] scoped rules
- [x] removing the Symbol type
- [x] use the block signature as revocation id
- [x] rename ID to Term
- [x] add an enum indicating the key's type
- [x] rename verifier to authorizer
- [x] update the parser to the latest grammar

TODO things to fix:
- [x] checking that pointer issue in protobuf for all other fields
- [x] checking that the deserialized keys are 32 byte arrays
- [ ] reintroduce the sig module: we'll need an abstraction over the key when we start supporting multiple algorithms
- [ ] verify signatures in Unmarshal: v1 was verifying the signatures while deserializing, and the Verify method was only checking that the root key is the expected one. Now Unmarshal only parses, while Verify checks the signature. In other implementations like [rust](https://docs.rs/biscuit-auth/latest/biscuit_auth/struct.Biscuit.html#method.from), we parse and check the signature in the same step, an creating the authorizer does not need to check anything (the rust lib also has  [UnverifiedBiscuit type](https://docs.rs/biscuit-auth/latest/biscuit_auth/struct.UnverifiedBiscuit.html), to be able to print or attenuate a token without checking the signature, it's a specific type that does not allow creating a verifier without checking the signature)
- [x] check that intersection and union are implemented
- [x] parse policies
- [ ] error management in the `ToExpr` parser methods (to convert from AST to builder structures)
- [x] remove quotes around `hex:...` byte arrays in the parser
- [ ] add tests for `.length()` (parsing and execution)
- [ ] add tests for intersection and union operations

### CODE REVIEW

**Comment by Geal** [comment]
signatures verify properly :partying_face: 

I'm encountering issues in [c0cad49](https://github.com/biscuit-auth/biscuit-go/pull/82/commits/c0cad497233beb220480fbfe7235bf6b7532bcc4) due to the way optional fields are serialized by protobuf. It uses pointers to represent them:

https://github.com/biscuit-auth/biscuit-go/pull/82/files#diff-c7507070ba25194410b4afae59dde5ab3743a0b1a23a33f87c4f5551cfa8756bR460

I tried using the methods provided by the proto package:
https://github.com/biscuit-auth/biscuit-go/pull/82/files#diff-75f279e5e9b923a88737f5ed1726cc2ef4f36e574c2d5a1ebf711df86c77e441R16

but it still serializes a pointer, not the value

**Comment by Geal** [comment]
now all tests pass. Next step: scoped rules

**Comment by Geal** [comment]
scoped rules are implemented, and I updated a bit the checks and rules printing. The last big chunk of the migration is removing the symbol type. The rest comes easily

**Comment by Geal** [comment]
the parser should be working now, but I still need to reactivate and convert all the parser tests from the old constraints to the new stack base expressions

**Comment by Geal** [comment]
I'm encountering an issue with the `.matches()` operation, where the parser apparently only sees `match` as a token, then `es` as a separate one

File: authorizer.go [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+func (v *authorizer) Authorize() error {
+	debug := datalog.SymbolDebugger{
+		SymbolTable: v.symbols,
+	}
+
+	// if we load facts from the verifier before
+	// the token's fact and rules, we might get inconsistent symbols
+	// token ements should first be converted to builder elements
```
Comment (daeMOn63) [inline authorizer.go:84]:
```suggestion
	// token elements should first be converted to builder elements
```

**Pull request merged**